### ISSUE TITLE
All these enums

### ISSUE THREAD

**Issue by joakim-hove** [OPM/opm-common#1471]
We have many enums in the OPM code .... some of them are only part of the opm implementation and can can hence be changed freelye - whereas others must be converted back and forth between eclipse specified integer values and opm enums; the conversion element is mostly relevant for restart files.

Currently the code *writing* restart files has many one way conversion functions of the type:
```
int fromOpm(OpmEnum e) {
...
} 
```
this has worked fine - but as part of #1396 there is a need to to go the other way as well. To simplify this I suggest writing one header file e.g. `compatible_enums.hpp` with functions `int to_ecl(T)` and `T from_ecl(int)` - something like this:
```
enum class OpmEnum1 {
     Value1,
     Value2,
     Value3
}

template <typename T>
T from_ecl(int ecl_value);

template <>
OpmEnum1 from_ecl(int ecl_value) {
      if (ecl_value == 17) 
          return OpmEnum::Value17;
      ...
}


int to_ecl(OpmEnum1 e) {
     if (e == OpmEnum::Value17)
           return 17;
      ...
}
```

Opinions?

**Comment by atgeirr** [comment]
Seems ok for me, but then I have not done a lot of work in this space.

**Comment by joakim-hove** [comment]
More input on this? If no one objects I will refactor in this direction.

**Comment by bska** [comment]
>  If no one objects I will refactor in this direction.

To be honest, I don't think there should be a single header/implementation for this.  I also don't think the template-base approach is the best here.  It gives too little context and I'm worried that the overload set will become too unwieldy.  Moreover, there are cases where the mapping isn't one-to-one.  The primary example would be the [WellType](https://github.com/OPM/opm-common/blob/13fa36f63b70cace2703d4e23f5281e17cf26340/src/opm/output/eclipse/LoadRestart.cpp#L1052-L1079) value.  These actually work really well as untyped `enum`s today and adding strong `enum`s would be a step backwards in usability.

On the other hand if we're going to add this kind of feature, then I'd much rather tie it to the types themselves rather than offer a veneer of generality. For instance, something along the lines of
```C++
struct WellType
{
    enum class Value {
        Unknown  = 0,
        Producer = 1,
        OilInj   = 2,
        WatInj   = 3,
        GasInj   = 4,
    };

    static int toInt(const Value v);
    static Value fromInt(const int v);
};

int WellType::toInt(const Value v)
{
    switch (v) {
    case Producer:  [[fallthrough]];
    case OilInj:    [[fallthrough]];
    case WatInj:    [[fallthrough]];
    case GasInj:
        return static_cast<int>(v);
    }

    throw std::invalid_argument {
        "Unsupported WellType Value: '" + std::to_string(static_cast<int>(v)) + '\''
    };
}

WellType::Value WellType::fromInt(const int v)
{
    switch (v) {
    case 1: return Value::Producer;
    case 2: return Value::OilInj;
    case 3: return Value::WatInj;
    case 4: return Value::GasInj:
    }

    return Value::Unknown;
}
```

**Comment by joakim-hove** [comment]
> The primary example would be the WellType value. These actually work really well as untyped enums today

yes that I agree - but the point is that I know need to go the other way - i.e. from an integer value, for `WellType` and many other enums.

> [...] then I'd much rather tie it to the types themselves rather

I agree that tying the functionality to the types this way looks nice - I have shied away from it as a bit of ever engineering - but it has really annoyed me that this is not immediately possible with `enum class` - the name should indicate that is possible.

I would prefer to have the enums define in the main header file they are used - i.e.  well related enums in `Well.hpp` - the reason I suggested having them in a separate header file was to be able to access them both from e.g. `Well.hpp` and `RestartIO::well.hpp` and also include `RestartIO::well.hpp` in `Well.hpp`.
